Update documentation in README

mpvosseller opened the pull request:
- Updates to latest syntax / api
- Updates to prefer enums that start with lowercase letters
- Added documentation on using the new `authorizationHeader` property

mpvosseller committed:
Update with latest syntax

mpvosseller committed:
Added documentation on using the authorizationHeader property

pietbrauer commented on README.md:
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
 public struct TokenConfiguration: Configuration {
     public let accessToken: String?
-    public let apiEndpoint = "https://my.webservice.example/api/2.0/"
+    public let apiEndpoint = "https://my.webservice.example/api/2.0/"    
```
Is the whitespace needed here?

pietbrauer commented on README.md:
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
 struct MyWebservice {
     var configuration: Configuration
-
+    
```
Please remove the whitespace

pietbrauer commented on README.md:
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
-    case GetMyself(Configuration)
-
+    case getMyself(Configuration)
+    
```
Please remove the extra whitespace.

pietbrauer commented on README.md:
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
         }
     }
-
+    
```
Please remove the extra whitespace.

pietbrauer commented on README.md:
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
         }
     }
-
+    
```
Please remove the extra whitespace.

pietbrauer commented on README.md:
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
         }
     }
-
+    
```
Please remove the extra whitespace.

pietbrauer commented on README.md:
```diff
@@ -33,38 +33,38 @@ This is what a basic router looks like:
     }
-
-    var params: [String: AnyObject] {
+    
```
Please remove the extra whitespace.

pietbrauer commented on README.md:
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
+public struct TokenConfiguration: Configuration {
+    public let accessToken: String?
+    public let apiEndpoint = "https://my.webservice.example/api/2.0/"
+    public let authorizationHeader: String? = "Bearer"
```
We can omit the `: String?` here.

mpvosseller replied:
Frustratingly it appears that we can't omit the type here. I'm not sure if this is the expected Swift behavior or if I'm doing something dumb or if this is a Swift bug? I put together some code you can throw into a playground to demo the issue:
https://gist.github.com/mpvosseller/02c6bbf853442c5de3c2fc9adc139e00

pietbrauer commented on README.md:
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
+    public let accessToken: String?
+    public let apiEndpoint = "https://my.webservice.example/api/2.0/"
+    public let authorizationHeader: String? = "Bearer"
+    public let errorDomain: String = "com.my.customErrorDomain"
```
We can omit the `: String` here.

pietbrauer commented on README.md:
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
         self.accessToken = accessToken
     }
 }
+
```
Please remove the extra whitespace.

pietbrauer commented on README.md:
````diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
+
 ```
 
+
````
Please remove the extra whitespace.

pietbrauer commented on README.md:
```diff
@@ -78,37 +78,56 @@ As RequestKit was designed to handle OAuth requests we needed something to store
-    func getMyself(session: RequestKitURLSession = NSURLSession.sharedSession(), completion: (response: Response<[String: AnyObject]>) -> Void) -> RequestKitURLSession {
-        let router = MyRouter.GetMyself(configuration)
-        router.loadJSON(session, expectedResultType: [String: AnyObject].self) { json, error in
+    
```
Please remove the extra whitespace.

pietbrauer reviewed (changes requested):
Nice work! I prefer no trailing whitespace but else this looks very very good.

mpvosseller committed:
fix whitespace

mpvosseller committed:
whitespace

mpvosseller committed:
whitespace

mpvosseller committed:
removed public

mpvosseller committed:
fix typo

mpvosseller committed:
removed unneeded type declaration

mpvosseller committed:
removed unneeded type declaration

mpvosseller commented:
OK I fixed the whitespace, removed a few unneeded type declarations, and also fixed a minor typo. 

The last remaining issue is that (shockingly) it does appear that an implementation of the `Configuration` protocol needs to explicitly declare `authorizationHeader` as an optional String in its implementation. 

So like this:
```
public let authorizationHeader: String? = "Bearer"
```
or like this
```
var authorizationHeader : String? {
    return "BEARER"
}
```

Unfortunately failing to do so results in the default implementation getting called which returns `nil`.

It is pretty annoying and I wonder if it is a Swift bug. Here is some code to demo the issue:
https://gist.github.com/mpvosseller/02c6bbf853442c5de3c2fc9adc139e00

If we are weary of doing this, the best alternative I can think of is to make this a non-optional String (with some default value like `BEARER`) that can be overridden and then add a new protocol method which is what we use to decide whether or not to use it like `shouldUseAuthorizationHeader`

WDYT?

mpvosseller commented:
After simplifying the code in this gist a bit I suspect this is probably the expected behavior from Swift.
https://gist.github.com/mpvosseller/02c6bbf853442c5de3c2fc9adc139e00

mpvosseller commented:
I had a conversation with Swift engineers on Twitter about this type issue. It appears to be a known issue which the may add compiler warnings for. I also filed a radar for the issue.
https://twitter.com/mpv/status/968938220225474560

The options I see are:

**Option 1** - Leave it as is and RequestKit users must explicitly declare their `authorizationHeader` implementation as `String?` and if they fail to do so the compiler (for now) won't complain and RequestKit just won't use the header. In this case we should update the docs to explain this.

**Option 2** - Make `authorizationHeader` a non optional `String` that defaults to `Bearer` and add a new protocol property like `shouldUseAuthorizationHeader` that returns a `Bool` and returns `false` by default.

WDYT? Any other options?

pietbrauer commented:
I would opt for Option 1
